### PR TITLE
x86: Tickless kernel and add missing CPU idle govs

### LINKED ISSUE #16313: x86/64: CPU never reaches minimum frequency (AMD Ryzen 7 5800U)
### Describe the bug

When booted into OpenWrt, the CPU freq never hits the minimum frequency of 400 MHz which it does under Arch Linux (dual booting on the same hardware).  Both distros run the 6.6.48 kernel and have the same scaling_governor settings:
```
# cat /sys/devices/system/cpu/cpufreq/policy0/scaling_governor
powersave
# cat /sys/devices/system/cpu/amd_pstate/status
active
# cat /sys/devices/system/cpu/cpufreq/policy0/scaling_driver
amd-pstate-epp
```

On Arch: 400 MHz
On OpenWrt: 1,384 MHz

### OpenWrt version

r27283-85f0128851

### OpenWrt release

SNAPSHOT

### OpenWrt target/subtarget

x86/64

### Device

HC Technology.,Ltd. HCAR5000-MI

### Image kind

Official downloaded image

### Steps to reproduce

1. Boot image
2. Confirm scaling governor, p-state driver (see above)
3. Run `cat /sys/devices/system/cpu/cpufreq/policy0/scaling_cur_freq`

### Actual behaviour

CPU never throttles down to the minimum frequency

### Expected behaviour

The CPU should idle at minimum frequency

### Additional info

I ever tried resetting to factory settings (stock `/etc/config/network`) and disabled all interfaces, and daemons and confirmed with `htop` that the CPU was as idle, but still the CPU would not throttle down to the minimum frequency.

Interestingly, if I `echo disable > /sys/devices/system/cpu/amd_pstate/status` I am able to hit the lowest frequency of 400 MHz under OpenWrt. I think that establishes two things:

- The processor is physically able to idle at 400 MHz under OpenWrt
- Something is likely wrong with the way the amd p-state governor operates under OpenWrt

### Diffconfig

_No response_

### Terms

- [X] I am reporting an issue for OpenWrt, not an unsupported fork.

### PR DESCRIPTION
**Pull request by graysky2**
Switch the x86 kernel's timer to tickless operation which is more power efficient since it is not woken up by periodic timer interrupts when idle.  Without this commit, my AMD Ryzen 7 5800U can only achieve a minimum core frequency of 1,384 MHz which is over 3x higher than the processor's minimum frequency of 400 MHz which is accessible with this modification.  In addition to the lower clock rate, I have seen a concomitant reduction in both idle temps and at-the-wall power consumption.

This commit fixes and closes #16313.

Summary:
* Idle CPU freqs dropped from 1,384 MHz to 400 Mhz.
* Idle power consumption dropped from 7 W avg to 5 W.
* Idle temps have dropped from 50C on avg to 43C.

There are other well known reasons to switch to a tickless timer including: reduced interrupt overhead, better use of CPU resources, and reduced latency to name a few.

This commit also builds in several other options for CPU idle governors.

Build system: x86/64
Build-tested: x86/64/AMD Cezanne
Run-tested: x86/64/AMD Cezanne

### CODE REVIEW

**Comment by graysky2** [comment]
I have found none running waveform's buffer bloat speed test or running iperf3.

```
% iperf3 -c 10.9.8.112
Connecting to host 10.9.8.112, port 5201
[  5] local 10.9.8.101 port 48046 connected to 10.9.8.112 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   283 MBytes  2.37 Gbits/sec    0    559 KBytes       
[  5]   1.00-2.00   sec   280 MBytes  2.35 Gbits/sec    0    617 KBytes       
[  5]   2.00-3.00   sec   281 MBytes  2.36 Gbits/sec    0    617 KBytes       
[  5]   3.00-4.00   sec   281 MBytes  2.35 Gbits/sec    0    617 KBytes       
[  5]   4.00-5.00   sec   280 MBytes  2.35 Gbits/sec    0    617 KBytes       
[  5]   5.00-6.00   sec   281 MBytes  2.36 Gbits/sec    0    617 KBytes       
[  5]   6.00-7.00   sec   281 MBytes  2.36 Gbits/sec    0    617 KBytes       
[  5]   7.00-8.00   sec   281 MBytes  2.36 Gbits/sec    0    648 KBytes       
[  5]   8.00-9.00   sec   281 MBytes  2.35 Gbits/sec    0    648 KBytes       
[  5]   9.00-10.00  sec   280 MBytes  2.35 Gbits/sec    0    648 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  2.74 GBytes  2.36 Gbits/sec    0             sender
[  5]   0.00-10.00  sec  2.74 GBytes  2.35 Gbits/sec                  receiver

iperf Done.
```

**Comment by graysky2** [comment]
Happy to test.  On the router with this PR applied, I can run `iperf3 -s -f m` and another machine, I will run `iperf3 -c 10.9.8.1` ... are you asking me to do that and then run `nft flush ruleset` and repeat the same test?

**Comment by vincejv** [comment]
sorry what i meant was to run this wireguard loopback test found in the forums, i think no need for flushing nft if the firewall config is not complex? i just read the post here in the forums..

https://forum.openwrt.org/t/a-wireguard-comparison-db/187586

> opkg update
> opkg install wireguard-tools ip-full kmod-veth iperf3 git-http
> git clone https://github.com/cyyself/wg-bench
> cd wg-bench
> ./setup-netns.sh
> ./benchmark.sh
> (You can also run in reverse way by ./benchmark.sh -R)
> After test, clean up by ./clean-up.sh

**Comment by graysky2** [comment]
I have a wireguard interface (`wg0`) already setup on the device.  I don't fully understand the scripts you linked.  Can you simplify the commands for me?

**Comment by vincejv** [comment]
you just have to run this in shell

1. `opkg update` to update the openwrt packages
2. `opkg install wireguard-tools ip-full kmod-veth iperf3 git-http` install dependencies for the benchmark
3. `git clone https://github.com/cyyself/wg-bench` clone the git benchmark repository in your openwrt device
4. `cd wg-bench` cd into the repo
5. `./setup-netns.sh` setup the wireguard loopback interface
6. `./benchmark.sh` run the benchmarking app
7. After test, clean up and remove the virtual interfaces by `./clean-up.sh`

anyways it's optional, up to you if you want to test, might do it on my machine too with this PR over the weekend

**Comment by graysky2** [comment]
I'm happy to test but I would rather type the commands than rely on those scripts.  If it just iperf3 going through the wg0 interface, I am asking if you know the minimum set of commands to type in order to do that.  I posted the same in the thread.

**Comment by schm0** [comment]
i guess it would be more interesting to see
how this behaves with low bandwith streams, like games... if this increases jitter

**Comment by vincejv** [comment]
iperf is reporting lower bandwidth on my i7-9500t machine when using the Wireguard benchmark scripts by cyyself, benchmark methodology is _somewhat_ explained here: https://github.com/cyyself/wg-bench/issues/35

**Before:**
```
# ./benchmark.sh
Connecting to host 169.254.200.2, port 5201
[  5] local 169.254.200.1 port 47378 connected to 169.254.200.2 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   866 MBytes  7.26 Gbits/sec    0    588 KBytes       
[  5]   1.00-2.00   sec   880 MBytes  7.38 Gbits/sec    0    652 KBytes       
[  5]   2.00-3.00   sec   884 MBytes  7.42 Gbits/sec    0    673 KBytes       
[  5]   3.00-4.00   sec   894 MBytes  7.50 Gbits/sec    0    644 KBytes       
[  5]   4.00-5.00   sec   899 MBytes  7.54 Gbits/sec    0    751 KBytes       
[  5]   5.00-6.00   sec   898 MBytes  7.53 Gbits/sec    0    729 KBytes       
[  5]   6.00-7.00   sec   895 MBytes  7.51 Gbits/sec    0    756 KBytes       
[  5]   7.00-8.00   sec   892 MBytes  7.47 Gbits/sec    0    727 KBytes       
[  5]   8.00-9.00   sec   885 MBytes  7.43 Gbits/sec    0    639 KBytes       
[  5]   9.00-10.00  sec   919 MBytes  7.70 Gbits/sec    0    508 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  8.70 GBytes  7.47 Gbits/sec    0             sender
[  5]   0.00-10.00  sec  8.70 GBytes  7.47 Gbits/sec                  receiver
```

**After this PR:**
```
# ./benchmark.sh 
Connecting to host 169.254.200.2, port 5201
[  5] local 169.254.200.1 port 22228 connected to 169.254.200.2 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   818 MBytes  6.86 Gbits/sec    0    716 KBytes       
[  5]   1.00-2.00   sec   842 MBytes  7.07 Gbits/sec    0    780 KBytes       
[  5]   2.00-3.00   sec   858 MBytes  7.20 Gbits/sec    0    671 KBytes       
[  5]   3.00-4.00   sec   864 MBytes  7.24 Gbits/sec    0    745 KBytes       
[  5]   4.00-5.00   sec   859 MBytes  7.21 Gbits/sec    0    647 KBytes       
[  5]   5.00-6.00   sec   860 MBytes  7.22 Gbits/sec    0    810 KBytes       
[  5]   6.00-7.00   sec   880 MBytes  7.38 Gbits/sec    0    866 KBytes       
[  5]   7.00-8.00   sec   866 MBytes  7.27 Gbits/sec    0    612 KBytes       
[  5]   8.00-9.00   sec   860 MBytes  7.22 Gbits/sec    0    676 KBytes       
[  5]   9.00-10.00  sec   868 MBytes  7.28 Gbits/sec    0    700 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  8.38 GBytes  7.20 Gbits/sec    0             sender
[  5]   0.00-10.00  sec  8.37 GBytes  7.19 Gbits/sec                  receiver
```

**Comment by vincejv** [comment]
yeah i repeated the tests again, **without the PR** is always producing more bandwidth...

**Comment by graysky2** [comment]
I don’t know what this test really proves or disproves. All we can see is about a 6% difference, but the throughout numbers are so high. It’s not clear to me what this is showing in a real world use case. I’m also wondering if it’s going to be CPU dependent. Will more powerful machine show less of a different or known at all?

**Comment by curtdept** [comment]
@vincejv 
I suspect the difference is packet coalescing in netdev, what offload options do you have on? 

If you look, you see the cwnd (congestion windows) are significantly higher in the PR versions.

**Comment by graysky2** [comment]
> I haven't tested this PR but I think it's preferred if CONFIG_HZ is set to 100 like it already is in the generic config. I'm not sure when it should be overriden.

Do you have a reference for that setting?  I understood that to be ignored despite being set.  I simple modeled the 300 setting after several other distro's config files.

**Comment by rsalvaterra** [comment]
> Do you have a reference for that setting? I understood that to be ignored despite being set. I simple modeled the 300 setting after several other distro's config files.

I also questioned this in my review (which hasn't yet been addressed, by the way). Scheduling three times more often per second comes at a (power) cost. Please don't do it if there's no measurable performance difference.

**Comment by vincejv** [comment]
> type the commands than rely on those scripts.

@graysky2 to answer your question

here are the commands to type to benchmark, comments to document what section you are in, high level explanation it's a wireguard loopback test, this creates a virtual wireguard interface for testing and run iperf locally on the same machine
```sh
# Create netns
ip netns add wg-bench

# Create veth
ip link add wg-bench type veth peer name wg-bench-ns
ip link set wg-bench-ns netns wg-bench
ip addr add 11.0.0.1/30 dev wg-bench
ip netns exec wg-bench ip addr add 11.0.0.2/30 dev wg-bench-ns
ip link set wg-bench up
ip netns exec wg-bench ip link set wg-bench-ns up

# Setup WireGuard for host
wg genkey > host-privkey 2> /dev/null
wg genkey > ns-privkey 2> /dev/null
ip link add wg-bench-wg type wireguard
wg set wg-bench-wg listen-port 11001 private-key host-privkey
ip addr add 169.254.200.1/32 dev wg-bench-wg peer 169.254.200.2
wg set wg-bench-wg peer $(wg pubkey < ns-privkey) allowed-ips 169.254.200.2/32 endpoint 11.0.0.2:11002
ip link set wg-bench-wg up

# Setup WireGuard for netns
ip netns exec wg-bench ip link add wg-bench-wg type wireguard
ip netns exec wg-bench wg set wg-bench-wg listen-port 11002 private-key $PWD/ns-privkey
ip netns exec wg-bench ip addr add 169.254.200.2/32 dev wg-bench-wg peer 169.254.200.1
ip netns exec wg-bench wg set wg-bench-wg peer $(wg pubkey < $PWD/host-privkey) allowed-ips 169.254.200.1/32 endpoint 11.0.0.1:11001
ip netns exec wg-bench ip link set wg-bench-wg up

# Run iperf3 benchmark
ip netns exec wg-bench iperf3 -s -D -p 4242
iperf3 -c 169.254.200.2 "$@" -p 4242
ip netns exec wg-bench fuser -k 4242/tcp
```

@curtdept
> what offload options do you have on?

not sure if it applies, since the script above creates a virtual interface and it's a "loopback" test, all traffic is generated within the same machine

**Comment by graysky2** [comment]
> I also questioned this in my review (which hasn't yet been addressed, by the way). Scheduling three times more often per second comes at a (power and throughput) cost. Please don't do it if there's no measurable performance difference.

I modified keeping the values of 100 in the generic config.

**Comment by curtdept** [comment]
@vincejv interesting, and this PR is the only diff relative to the non-PR version?

**Comment by vincejv** [comment]
@curtdept yes but i'm not sure if it's margin of error, I only cherry-picked this commit and compile, but i ran the test 4 times already rebooting, swapping kernels, but I still want others to check to verify

**Comment by rany2** [comment]
> I also questioned this in my review (which hasn't yet been addressed, by the way).

I can't find your review though, did you submit it?

**Comment by rsalvaterra** [comment]
> I can't find your review though, did you submit it?

Yes, but seems it can only be seen when logged in…? I admit not being very GitHub-literate. To me it shows like this…
Screenshot at 2024-09-07 10-01-07

**Comment by graysky2** [comment]
For some reason it shows from your screenshot as "pending" ... I too was not notified that you started a review nor can I see it.

**Review by rsalvaterra — CHANGES_REQUESTED** [review]

**Comment by rsalvaterra** [comment]
> For some reason it shows from your screenshot as "pending" ... I too was not notified that you started a review nor can I see it.

Yeah, my bad, I'm not familiar with the workflow. Should be visible now. :slightly_smiling_face:

File: target/linux/x86/config-6.6 [hunk 1]
```diff
@@ -179,7 +183,12 @@ CONFIG_HW_RANDOM=y
 CONFIG_HW_RANDOM_GEODE=y
 CONFIG_HW_RANDOM_VIA=y
 # CONFIG_HYPERVISOR_GUEST is not set
-CONFIG_HZ_PERIODIC=y
+CONFIG_HZ_300=y
+CONFIG_HZ=300
+# CONFIG_HZ_PERIODIC is not set
+CONFIG_NO_HZ=y
+CONFIG_NO_HZ_COMMON=y
+CONFIG_NO_HZ_FULL=y
```
[the remark this answers is not in the document]
Comment (graysky2) [reply target/linux/x86/config-6.6:191]: Of the three options, are you recommending `CONFIG_NO_HZ_IDLE=y`?

From https://www.kernel.org/doc/html/latest/timers/no_hz.html:


> There are three main ways of managing scheduling-clock interrupts (also known as “scheduling-clock ticks” or simply “ticks”):
> 
> Never omit scheduling-clock ticks (CONFIG_HZ_PERIODIC=y or CONFIG_NO_HZ=n for older kernels). You normally will -not- want to choose this option.
> 
> Omit scheduling-clock ticks on idle CPUs (CONFIG_NO_HZ_IDLE=y or CONFIG_NO_HZ=y for older kernels). This is the most common approach, and should be the default.
> 
> Omit scheduling-clock ticks on CPUs that are either idle or that have only one runnable task (CONFIG_NO_HZ_FULL=y). Unless you are running realtime applications or certain types of HPC workloads, you will normally -not- want this option.
>

[the remark this answers is not in the document]
Comment (rsalvaterra) [reply target/linux/x86/config-6.6]: Yes, that was the idea, thanks!

**Comment by rsalvaterra** [comment]
While it looks much better to me, I'd do an additional test comparing `cpuidle`/`cpufreq` governors. More specifically, the combination of `menu`/`ondemand` (traditional) with `teo`/`schedutil` (more recent).

**Comment by graysky2** [comment]
Quite a bit of work has been merged with the `amd-pstate-epp` and `amd-pstate` scaling_driver.  I have experimented with it and with values of `active`, `guided`, and `passive` and concluded that they are all so similar (differing by 10th of a watt and I don't even trust the meter to that level of precision that I just went with the default of `amd-pstate-epp`/`active` .

Currently:
* `/sys/module/cpufreq/parameters/default_governor` is `schedutil`
* `/sys/devices/system/cpu/cpuidle/current_driver` is `acpi_idle`
* `/sys/devices/system/cpu/cpuidle/current_governor` is `menu`

I need to read about these to understand what they do and why I would want to change them.

File: target/linux/x86/config-6.6 (hunk 1 above)
[the remark this answers is not in the document]
Comment (graysky2) [reply target/linux/x86/config-6.6]: @rsalvaterra - I reopened this after reading some more.  I think changing this back to `CONFIG_NO_HZ_FULL=y` is more flexible for users.  While it is true that users will need to specify boot parameters (see below), without them defined, the system will operate as if `NO_HZ_IDLE` is enabled.

* `nohz_full=` to define which cores will be in full tickless mode
* `isolcpus=` to define which cores are isolated for user tasks
* `rcu_nocbs=` to offload RCU (Read-Copy-Update) callbacks from the CPUs in full tickless mode, further reducing the load on them.

What do you think?

[the remark this answers is not in the document]
Comment (rsalvaterra) [reply target/linux/x86/config-6.6]: Well, even though I presonally don't think this is useful in OpenWrt, there's probably someone out there crazy enough to prove me wrong. And x86-64-based systems aren't exactly storage/RAM/CPU-constrained. I guess it's fine to enable `CONFIG_NO_HZ_FULL` by default.

[the remark this answers is not in the document]
Comment (graysky2) [reply target/linux/x86/config-6.6:190]: > there's probably someone out there crazy enough to prove me wrong

Yes!  I'm sure someone out there would like OpenWrt support for their kerosene-powered cheese grater.

**Comment by vincele** [comment]
I'd do this as two distinct commits (or even PRs) because it is mixing 2 different things:
* tickless
* adding governors

They may look related, but I don't think that's a valid reason to put both in at the same time.

And you can then get performance numbers for them separately.

Just my 2 cents.

**Comment by graysky2** [comment]
Good idea, @vincele, done

**Comment by vincele** [comment]
This is better, but you kept the previous numbers in the tickless commit message. I think they should be reassessed separately for both commits. Or were they from only the tickless change ? Or were they with the modified HZ, as @rany2 was alluding to in [his comment](https://github.com/openwrt/openwrt/pull/16317#issuecomment-2334512590) ?

**Comment by graysky2** [comment]
Actually, I think this PR is better as a single commit for the following reasons:
1. It would be incomplete without the addition of upstream's default CPU idle gov. (menu)
2. It will make any future bisect more straightforward

**Comment by graysky2** [comment]
@rsalvaterra - if you're happy with PR as-is can you please clear your change request?
yes

**Comment by rany2** [comment]
Some config entries should be sorted properly. Also, others could be dropped entirely as they are already defined in the generic config.

```
# ./scripts/kconfig.pl - '>' target/linux/{generic,x86}/config-6.6 target/linux/generic/config-filter > target/linux/x86/config-6.6.new
# mv target/linux/x86/config-6.6{.new,}
# git diff | cat
diff --git a/target/linux/x86/config-6.6 b/target/linux/x86/config-6.6
index ddd3479ee1..c7fd528395 100644
--- a/target/linux/x86/config-6.6
+++ b/target/linux/x86/config-6.6
@@ -183,13 +183,6 @@ CONFIG_HW_RANDOM=y
 CONFIG_HW_RANDOM_GEODE=y
 CONFIG_HW_RANDOM_VIA=y
 # CONFIG_HYPERVISOR_GUEST is not set
-CONFIG_HZ_100=y
-CONFIG_HZ=100
-CONFIG_NO_HZ=y
-CONFIG_NO_HZ_COMMON=y
-CONFIG_NO_HZ_FULL=y
-# CONFIG_NO_HZ_IDLE is not set
-# CONFIG_HZ_PERIODIC is not set
 CONFIG_I8253_LOCK=y
 CONFIG_IA32_FEAT_CTL=y
 # CONFIG_IB700_WDT is not set
@@ -284,6 +277,9 @@ CONFIG_NET_XGRESS=y
 CONFIG_NLS=y
 # CONFIG_NMI_CHECK_CPU is not set
 # CONFIG_NOHIGHMEM is not set
+CONFIG_NO_HZ=y
+CONFIG_NO_HZ_COMMON=y
+CONFIG_NO_HZ_FULL=y
 CONFIG_NR_CPUS=1
 CONFIG_NR_CPUS_DEFAULT=1
 CONFIG_NR_CPUS_RANGE_BEGIN=1
```

**Comment by graysky2** [comment]
Thanks @rany2 - I did not use that script before.

**Comment by vincejv** [comment]
Is tickless preferred over Realtime kernel (PREEMPT_RT on mainline linux 6.12) for OpenWrt devices?

**Comment by rsalvaterra** [comment]
> Is tickless preferred over Realtime kernel (PREEMPT_RT on mainline linux 6.12) for OpenWrt devices?

PREEMPT_RT is useless for OpenWrt. Rule of thumb, if you don't know if you need PREEMPT_RT or not, you don't need it.

**Review by rsalvaterra — APPROVED** [review]
LGTM now, thanks (and sorry for the late reply).

**Comment by Ansuel** [comment]
@graysky2 merging this but I would love some effort in making selecting idle gov consistent and selectable in menuconfig... (and filter those option...)

File: target/linux/x86/config-6.6 [hunk 2]
```diff
@@ -275,6 +278,9 @@ CONFIG_NET_XGRESS=y
 CONFIG_NLS=y
 # CONFIG_NMI_CHECK_CPU is not set
 # CONFIG_NOHIGHMEM is not set
+CONFIG_NO_HZ=y
+CONFIG_NO_HZ_COMMON=y
+CONFIG_NO_HZ_FULL=y
```
Comment (brada4) [inline target/linux/x86/config-6.6:283]: I dont think anyone will play with boot parameters. No gain in normal case.
https://jeremyeder.com/2013/11/15/nohz_fullgodmode/

**Comment by jisakiel** [comment]
It would [appear](https://github.com/openwrt/openwrt/blob/openwrt-25.12/target/linux/x86/config-6.12#L187) these commits were lost unadvertently by a refresh https://github.com/openwrt/openwrt/commit/805b5d893c424719d845d2840bdd3e41932582f5 ?

EDIT: nvm, NO_HZ seems there even if some settings were dropped along the way https://github.com/openwrt/openwrt/blame/main/target/linux/x86/config-6.12#L326 . My bad!

**Pull request merged**